Handle allOf with optional

mstriemer opened the pull request:
This handles parsing schemas with an `allOf` that has an optional schema in it. This was causing an error on `content_scripts` without a `run_at` property.

Fixes #1245.

tofumatt commented on tests/schema/test.firefox-schemas-import.js:
```diff
@@ -82,6 +82,50 @@ describe('firefox schema import', () => {
+      };
+      assert.deepEqual(rewriteOptionalToRequired(obj), {
+        foo: { type: 'string' },
+        bar: { allOf: [
```
Nitpick but it would be nice if this followed the layout the same as above to visually compare them.

tofumatt commented on tests/schema/test.firefox-schemas-import.js:
```diff
@@ -82,6 +82,50 @@ describe('firefox schema import', () => {
       });
     });
+
+    it('handles an allOf with one being optional', () => {
```
I think you should say "with one property being optional" or something like that because this isn't really clear.

tofumatt commented on tests/schema/test.firefox-schemas-import.js:
```diff
@@ -82,6 +82,50 @@ describe('firefox schema import', () => {
+      });
+    });
+
+    it('handles an allOf with none being optional', () => {
```
Same here; this should be "with no property flagged as optional" or something like that.

tofumatt commented on src/schema/firefox-schemas-import.js:
```diff
@@ -68,7 +68,18 @@ export function rewriteOptionalToRequired(schema) {
     if (!Array.isArray(value) && typeof value === 'object') {
       const { optional, ...rest } = value;
-      if (!optional) {
+      if (Array.isArray(rest.allOf)) {
```
This is a bit of an odd to read chunk of code. I get it after looking at the issue, but I wouldn't mind a link back to the issue just to explain the reasoning behind it.

mstriemer commented:
I had added a test to reproduce the bug earlier and forgot to add it to git. It's added now. I gave it a once over and it looks okay but let me know if something is amiss and I can update it.

I also refactored the `allOf.map` loop to use destructuring a bit better and avoid deleting the property explicitly. I think it makes it a bit simpler. Added some comments too so it should be easier to understand now.

tofumatt commented:
Looks good to me 👍
